### PR TITLE
endpoint to persist "Request access" form submissions

### PR DESCRIPTION
**Pull request by the-daniel-rothig**
New configs required
email:host
email:user
email:password

... for access to an smtp server

### CODE REVIEW

**Comment by the-daniel-rothig** [comment]
Note that clients compiling swagger.json now need to specify a new flag, see commit message of this commit: https://github.com/DFE-Digital/manage-courses-ui/commit/144ba159cceeaf7d055690f32a217d21d6e42a32

File: src/ManageCourses.Api/Controllers/AccessRequestController.cs [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        public async Task<StatusCodeResult> Index([FromForm] Api.Model.AccessRequest request) 
+        {
+            var requesterEmail = "daniel@eigencode.io"; //this.User.Identity.Name;
+            using (var transaction = ((DbContext)_context).Database.BeginTransaction()) 
+            {
+                try
+                {
+                    var requester = _context.McUsers
+                        .Include(x=>x.McOrganisationUsers)
+                        .ThenInclude(x=>x.McOrganisation).Single(x => x.Email == requesterEmail);
```
Comment (defong) [inline src/ManageCourses.Api/Controllers/AccessRequestController.cs]: Are we not going to use lazy loading?
I was halfway through to setting up entity framework to use that....

File: src/ManageCourses.Api/Controllers/AccessRequestController.cs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+                        LastName = request.LastName,
+                        EmailAddress = request.EmailAddress,
+                        Organisation = request.Organisation,
+                        Reason = request.Reason,
+                        Status = Domain.Models.AccessRequest.RequestStatus.Requested
+                    }); 
+                    _context.Save();
+
+                    var requestedBody = "The recipient is not yet on the system.\n\n";
+                    if (requestedIfExists != null) {
```
Comment (defong) [inline src/ManageCourses.Api/Controllers/AccessRequestController.cs]: Can we change the variable to reflect that's its a bool?

File: src/ManageCourses.Api/Controllers/AccessRequestController.cs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+                        $"Email address: {request.EmailAddress}\n" +
+                        $"Organisation address: {request.Organisation}\n" +
+                        $"Reason: {request.Reason}\n\n" +
+
+                        requestedBody +
+
+                        $"The recipient would gain access to the following organisation(s):\n" +
+                        string.Join("\n", orgs.Select(x => $"- {x}\n").ToArray());
+
+                    System.Console.WriteLine(emailBody);
```
Comment (defong) [inline src/ManageCourses.Api/Controllers/AccessRequestController.cs]: Why? Use logger instead.

File: src/ManageCourses.Api/Controllers/AccessRequestController.cs [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        {
+            _context = context;
+            _emailService = emailService;
+        }
+
+        //[Authorize]
+        [HttpPost]
+        public async Task<StatusCodeResult> Index([FromForm] Api.Model.AccessRequest request) 
+        {
+            var requesterEmail = "daniel@eigencode.io"; //this.User.Identity.Name;
```
Comment (defong) [inline src/ManageCourses.Api/Controllers/AccessRequestController.cs]: Why?

File: src/ManageCourses.Api/Data/EmailService.cs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        private SmtpClient _smtpClient;
+        private string _user;
+
+        public EmailService(string host, string user, string password)
+        {
+            _user = user;
+            _smtpClient = new SmtpClient
+            {
+                Host = host, // set your SMTP server name here
+                Port = 587, // Port
```
Comment (defong) [inline src/ManageCourses.Api/Data/EmailService.cs]: Why hardcoded?

File: src/ManageCourses.Api/Controllers/AccessRequestController.cs (hunk 1 above)
Comment (the-daniel-rothig) [reply src/ManageCourses.Api/Controllers/AccessRequestController.cs]: Don't need lazy loading here - all elements in scope need to be retrieved. The filter (email == email) will be evaluated server-side, thanks to EF magic

File: src/ManageCourses.Api/Data/EmailService.cs (hunk 5 above)
Comment (the-daniel-rothig) [reply src/ManageCourses.Api/Data/EmailService.cs]: I was on the fence about making this a parameter but I want this class to only support port 587, to be RFC compliant. I extracted this as a const to make this choice more explicit

File: src/ManageCourses.Api/Controllers/AccessRequestController.cs (hunk 4 above)
Comment (the-daniel-rothig) [reply src/ManageCourses.Api/Controllers/AccessRequestController.cs]: oops, leftover from debugging

File: src/ManageCourses.Api/Controllers/AccessRequestController.cs (hunk 3 above)
Comment (the-daniel-rothig) [reply src/ManageCourses.Api/Controllers/AccessRequestController.cs]: oops, leftover from debugging

File: src/ManageCourses.Api/Controllers/AccessRequestController.cs (hunk 2 above)
Comment (the-daniel-rothig) [reply src/ManageCourses.Api/Controllers/AccessRequestController.cs]: thanks - made logic more explicit by introducing bool

File: src/ManageCourses.Api/Data/EmailService.cs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+        public void SendEmailToSupport(string subject, string message)
+        {
+
+            var mailMessage = new MailMessage(_user, _user)
+            {
+                Subject = subject,
+                Body = message
+            };
+            {
```
Comment (defong) [inline src/ManageCourses.Api/Data/EmailService.cs]: Anything special with the braces?

File: src/ManageCourses.Api/Controllers/AccessRequestController.cs [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+                    }); 
+                    _context.Save();
+
+                    string requestedBody;
+                    bool requestedAlreadyExistsAsUser = requestedIfExists == null;
+                    if (requestedAlreadyExistsAsUser) 
+                    {                        
+                        requestedBody = 
+                            $"The recipient is already on the system as:\n"+
+                            $"First name: {requestedIfExists.FirstName}\n"+
```
Comment (defong) [inline src/ManageCourses.Api/Controllers/AccessRequestController.cs]: Ummm that will be null right?

File: README.md [hunk 8]
```diff
@@ -22,7 +22,7 @@ The project follows https://semver.org/ version numbering.
 
 In a windows command prompt:
 
-    cd src\api
+    cd src\ManageCourses.Api
```
Comment (dankmitchell) [inline README.md]: Should be `cd src/ManageCourses.Api`

**Comment by defong** [comment]
can you also add it to appsettings, as example etc...

"email": {
    "host": "host",
    "user": "email",
    "password": "password"
  }

File: src/ManageCourses.Api/Data/EmailService.cs (hunk 6 above)
Comment (the-daniel-rothig) [reply src/ManageCourses.Api/Data/EmailService.cs]: oops! thanks

File: src/ManageCourses.Api/Controllers/AccessRequestController.cs (hunk 7 above)
Comment (the-daniel-rothig) [reply src/ManageCourses.Api/Controllers/AccessRequestController.cs]: fixed - thanks for spotting

**Pull request merged**